### PR TITLE
MAINT: Separate correct `longdouble.__float__` from incorrect `longdouble.__int__`

### PR DESCRIPTION
**Pull request by eric-wieser**
These barely shared any useful code at all, and it makes it clearer where the fix needs to go for #9964

### CODE REVIEW

File: numpy/core/src/multiarray/scalartypes.c.src [hunk 1]
```diff
@@ -816,66 +816,54 @@ halftype_@kind@(PyObject *self)
...
-@char@longdoubletype_@name@(PyObject *self)
+@char@longdoubletype_float(PyObject *self)
 {
-    double dval;
-    PyObject *obj, *ret;
+    npy_longdouble val = PyArrayScalar_VAL(self, @CHAR@LongDouble)@POST@;
+    return PyFloat_FromDouble((double) val);
+}
 
-    dval = (double)(((Py@CHAR@LongDoubleScalarObject *)self)->obval)@POST@;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/scalartypes.c.src]: Using `PyArrayScalar_VAL` in favor of this

File: numpy/core/src/multiarray/scalartypes.c.src [hunk 2]
```diff
@@ -816,66 +816,54 @@ halftype_@kind@(PyObject *self)
...
- *
- * #name = (int, float)*2#
- * #KIND = (Long, Float)*2#
- * #char = ,,c*2#
- * #CHAR = ,,C*2#
- * #POST = ,,.real*2#
+ * #char = ,c#
+ * #CHAR = ,C#
+ * #POST = ,.real#
  */
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/scalartypes.c.src]: This repeat is way clearer expressed as two repeats

File: numpy/core/src/multiarray/scalartypes.c.src [hunk 3]
```diff
@@ -816,66 +816,54 @@ halftype_@kind@(PyObject *self)
...
+{
+    npy_longdouble val = PyArrayScalar_VAL(self, @CHAR@LongDouble)@POST@;
+
+    /*
+     * This raises OverflowError when the cast overflows to infinity (gh-9964)
+     *
+     * Could fix this with a PyLong_FromLongDouble(longdouble ldval)
+     * but this would need some more work...
+     */
+    return PyLong_FromDouble((double) val);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/scalartypes.c.src]: This is where things need to change to fix #9964, but lets leave that to a future PR.

File: numpy/core/src/multiarray/scalartypes.c.src [hunk 4]
```diff
@@ -4111,37 +4099,32 @@ initialize_numeric_types(void)
...
-    /**begin repeat
-     * #name = long, hex, oct#
-     * #kind = tp_as_number->nb*3#
+    /*
+     * These need to be coded specially because longdouble/clongdouble getitem
+     * does not return a normal Python type
      */
+    @char@longdoubletype_as_number.nb_float = @char@longdoubletype_float;
+#if defined(NPY_PY3K)
+    @char@longdoubletype_as_number.nb_int  = @char@longdoubletype_long;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/scalartypes.c.src]: This is deliberate, as nb_int should try to return `int` on py2, but `long` on py3. I'm not sure it makes a difference, but it's more consistent with how `longobject` is implemented.

File: numpy/core/src/multiarray/scalartypes.c.src [hunk 5]
```diff
@@ -4111,37 +4099,32 @@ initialize_numeric_types(void)
...
+    @char@longdoubletype_as_number.nb_long = @char@longdoubletype_long;
+    @char@longdoubletype_as_number.nb_hex  = @char@longdoubletype_hex;
+    @char@longdoubletype_as_number.nb_oct  = @char@longdoubletype_oct;
+#endif
 
-    PyLongDoubleArrType_Type.@kind@_@name@ = longdoubletype_@name@;
-    PyCLongDoubleArrType_Type.@kind@_@name@ = clongdoubletype_@name@;
+    Py@CHAR@LongDoubleArrType_Type.tp_as_number = &@char@longdoubletype_as_number;
+    Py@CHAR@LongDoubleArrType_Type.tp_repr = @char@longdoubletype_repr;
+    Py@CHAR@LongDoubleArrType_Type.tp_str = @char@longdoubletype_str;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/scalartypes.c.src]: It's far more readable and concise to iterate over the types here, rather than the members.

**Comment by charris** [comment]
Thanks Eric.

**Pull request merged**